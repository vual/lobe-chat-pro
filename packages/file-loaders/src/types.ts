// Define supported file types - consider using an enum or const assertion
export type SupportedFileType = 'pdf' | 'docx' | 'txt' | 'excel' | 'pptx'; // | 'pptx' | 'latex' | 'epub' | 'code' | 'markdown';

/**
 * 代表一个完整的已加载文件，包含文件级信息和其所有页面/块。
 */
export interface FileDocument {
  /**
   * 文件内容
   */
  content: string;

  /**
   * 文件创建时间戳。
   */
  createdTime: Date;

  /**
   * 文件类型或扩展名。
   */
  fileType: string;

  /**
   * 原始文件名。
   */
  filename: string;

  /**
   * 文件级别的元数据。
   * 例如从文件属性中提取的标题、作者，或整个文件加载失败时的错误。
   */
  metadata: {
    /**
     * 允许添加其他文件级别的元数据。
     */
    [key: string]: any;
    /**
     * 文档作者 (如果可用)。
     */
    author?: string;
    /**
     * 如果整个文件加载失败，记录错误信息。
     */
    error?: string;
    /**
     * 文档标题 (如果可用)。
     */
    title?: string;
  };

  /**
   * 文件最后修改时间戳。
   */
  modifiedTime: Date;

  /**
   * 包含文档中所有逻辑页面/块的数组。
   * 顺序通常对应文件中的自然顺序。
   */
  pages?: DocumentPage[];

  /**
   * 原始文件的完整路径。
   */
  source: string;

  /**
   * 整个文档的总字符数 (所有 Page 的 charCount 之和)。
   * 需要在所有 Page 加载和计算后得出。
   */
  totalCharCount: number;

  /**
   * 整个文档的总行数 (所有 Page 的 lineCount 之和)。
   * 需要在所有 Page 加载和计算后得出。
   */
  totalLineCount: number;
}

/**
 * 代表文件中的一个逻辑单元/页面/块。
 */
export interface DocumentPage {
  /**
   * 此页/块内容的字符数。
   */
  charCount: number;

  /**
   * 此页/块内容的行数。
   */
  lineCount: number;

  /**
   * 与此页/块相关的元数据。
   */
  metadata: {
    /**
     * 允许添加其他特定于页/块的元数据。
     */
    [key: string]: any;

    /**
     * 如果原始文件单元被进一步分割成块，这是当前块的索引。
     */
    chunkIndex?: number;

    /**
     * 处理此页/块时发生的错误。
     */
    error?: string;

    /**
     * 此页/块在原始文件中的结束行号。
     */
    lineNumberEnd?: number;

    /**
     * 此页/块在原始文件中的起始行号。
     */
    lineNumberStart?: number;

    /**
     * 页码 (适用于 PDF, DOCX)。
     */
    pageNumber?: number;

    /**
     * 与此页/块相关的章节标题。
     */
    sectionTitle?: string;

    /**
     * 工作表名称 (适用于 XLSX)。
     */
    sheetName?: string;

    /**
     * 幻灯片编号 (适用于 PPTX)。
     */
    slideNumber?: number;

    /**
     * 如果原始文件单元被进一步分割成块，这是该单元的总块数。
     */
    totalChunks?: number;
  };

  /**
   * 此页/块的核心文本内容。
   */
  pageContent: string;
}

/**
 * 可选的文件元数据，用于覆盖从文件系统读取的信息。
 */
export interface FileMetadata {
  /**
   * 文件创建时间戳。
   */
  createdTime?: Date;
  /**
   * 文件类型或扩展名。
   */
  fileType?: string;
  /**
   * 文件名。
   */
  filename?: string;
  /**
   * 文件最后修改时间戳。
   */
  modifiedTime?: Date;
  /**
   * 文件来源标识 (例如 S3 URL 或原始路径)。
   */
  source?: string;
}

/**
 * 定义所有文件加载器类必须实现的接口。
 */
export interface FileLoaderInterface {
  /**
   * 将从 loadPages 获取的页面内容聚合成单一的字符串。
   * @param pages DocumentPage 对象的数组。
   * @returns 返回聚合后的文本内容的 Promise。
   */
  aggregateContent(pages: DocumentPage[]): Promise<string>;

  attachDocumentMetadata?(filePath: string): Promise<Record<string, any>>;

  /**
   * 根据文件路径加载文件内容，并将其分割为逻辑页面/块。
   * @param filePath 文件的完整路径。
   * @returns 返回包含 DocumentPage 对象的数组的 Promise。
   */
  loadPages(filePath: string): Promise<DocumentPage[]>;
}
