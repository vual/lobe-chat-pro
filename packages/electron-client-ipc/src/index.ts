export * from './dispatch';
export * from './events';
export * from './types';
export * from './useWatchBroadcast';
