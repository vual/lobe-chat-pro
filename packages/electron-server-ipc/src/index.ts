export * from './events';
export * from './ipcClient';
export * from './ipcServer';
export * from './types';
