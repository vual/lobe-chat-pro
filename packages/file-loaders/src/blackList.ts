// List of system files/directories to ignore
export const SYSTEM_FILES_TO_IGNORE = [
  '.DS_Store',
  'Thumbs.db',
  'desktop.ini',
  '.localized',
  'ehthumbs.db',
  'ehthumbs_vista.db',
];
