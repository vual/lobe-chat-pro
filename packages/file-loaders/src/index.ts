export * from './blackList';
export * from './loadFile';
export * from './types';
