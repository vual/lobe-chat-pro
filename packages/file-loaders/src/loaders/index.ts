import { FileLoaderInterface, SupportedFileType } from '../types';
import { DocxLoader } from './docx';
// import { EpubLoader } from './epub';
import { ExcelLoader } from './excel';
import { PdfLoader } from './pdf';
import { PptxLoader } from './pptx';
import { TextLoader } from './text';

// Loader configuration map
// Key: file extension (lowercase, without leading dot) or specific type name
// Value: Loader Class implementing FileLoaderInterface
export const fileLoaders: Record<SupportedFileType, new () => FileLoaderInterface> = {
  docx: DocxLoader,
  // epub: EpubLoader,
  excel: ExcelLoader,
  pdf: PdfLoader,
  pptx: PptxLoader,
  txt: TextLoader,
};
