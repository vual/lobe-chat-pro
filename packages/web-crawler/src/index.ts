export type { CrawlImplType } from './crawImpl';
export { Crawler } from './crawler';
export * from './type';
