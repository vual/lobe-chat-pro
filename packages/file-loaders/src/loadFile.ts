import debug from 'debug';
import { stat } from 'node:fs/promises';
import * as path from 'node:path';

import { fileLoaders } from './loaders';
import { TextLoader } from './loaders/text';
import { FileDocument, FileMetadata, SupportedFileType } from './types';
import type { DocumentPage, FileLoaderInterface } from './types';
import { isTextReadableFile } from './utils/isTextReadableFile';

const log = debug('file-loaders:loadFile');

/**
 * Determines the file type based on the filename extension.
 * @param filePath The path to the file.
 * @returns The determined file type or 'txt' if text-readable, undefined otherwise.
 */
const getFileType = (filePath: string): SupportedFileType | undefined => {
  log('Determining file type for:', filePath);
  const extension = path.extname(filePath).toLowerCase().replace('.', '');

  if (!extension) {
    log('No extension found, treating as txt');
    return 'txt'; // Treat files without extension as text?
  }

  // Prioritize checking if it's a generally text-readable type
  if (isTextReadableFile(extension)) {
    log(`Extension '${extension}' is text-readable, treating as txt`);
    return 'txt';
  }

  // Handle specific non-text or complex types
  log(`Checking specific types for extension: '${extension}'`);
  switch (extension) {
    case 'pdf': {
      log('File type identified as pdf');
      return 'pdf';
    }
    case 'docx': {
      log('File type identified as docx');
      return 'docx';
    }
    case 'xlsx':
    case 'xls': {
      log('File type identified as excel');
      return 'excel';
    }
    case 'pptx': {
      log('File type identified as pptx');
      return 'pptx';
    }
    default: {
      log(
        `Extension '${extension}' is not a specifically handled type and not text-readable. Unsupported.`,
      );
      // If not text-readable and not a specific known type, it's unsupported
      return undefined;
    }
  }
};

// Default fallback loader class
const DefaultLoader = TextLoader;

/**
 * Loads a file from the specified path, automatically detecting the file type
 * and using the appropriate loader class.
 *
 * @param filePath The path to the file to load.
 * @param fileMetadata Optional metadata to override information read from the filesystem.
 * @returns A Promise resolving to a FileDocument object.
 */
export const loadFile = async (
  filePath: string,
  fileMetadata?: FileMetadata,
): Promise<FileDocument> => {
  log('Starting to load file:', filePath, 'with metadata:', fileMetadata);
  let stats;
  let fsError: string | undefined;

  try {
    log('Attempting to get file stats for:', filePath);
    stats = await stat(filePath);
    log('Successfully retrieved file stats:', stats);
  } catch (e) {
    const error = e as Error;
    log('Error getting file stats for %s: %s', filePath, error.message);
    console.error(`Error getting file stats for ${filePath}: ${error.message}`);
    fsError = `Failed to access file stats: ${error.message}`;
  }

  // Determine base file info from path and stats (if available)
  log('Determining base file info');
  const fileExtension = path.extname(filePath).slice(1).toLowerCase();
  const baseFilename = path.basename(filePath);

  // Apply overrides from fileMetadata or use defaults
  const source = fileMetadata?.source ?? filePath;
  const filename = fileMetadata?.filename ?? baseFilename;
  const fileType = fileMetadata?.fileType ?? fileExtension;
  const createdTime = fileMetadata?.createdTime ?? stats?.ctime ?? new Date();
  const modifiedTime = fileMetadata?.modifiedTime ?? stats?.mtime ?? new Date();
  log('File info determined/overridden: %O', {
    createdTime,
    fileType,
    filename,
    modifiedTime,
    source,
  });

  const paserType = getFileType(filePath);
  log('Parser type determined as:', paserType);

  // Select the loader CLASS based on the determined fileType, fallback to DefaultLoader
  const LoaderClass: new () => FileLoaderInterface = paserType
    ? fileLoaders[paserType]
    : DefaultLoader;
  log('Selected loader class:', LoaderClass.name);

  if (!paserType) {
    console.warn(
      `No specific loader found for file type '${fileType}'. Using default loader (${DefaultLoader.name}) as fallback.`,
    );
  }

  let pages: DocumentPage[] = [];
  let aggregatedContent = '';
  let loaderError: string | undefined;
  let aggregationError: string | undefined;
  let metadataError: string | undefined;
  let loaderSpecificMetadata: any | undefined;

  // Instantiate the loader
  log('Instantiating loader:', LoaderClass.name);
  const loaderInstance = new LoaderClass();

  // If we couldn't even get stats, skip loader execution
  if (!fsError) {
    log('File stats available, proceeding with loader execution.');
    try {
      // 1. Load pages using the instance
      log('Loading pages with loader:', LoaderClass.name, 'for file:', filePath);
      pages = await loaderInstance.loadPages(filePath);
      log('Pages loaded successfully, count:', pages.length);

      try {
        // 2. Aggregate content using the instance
        log('Aggregating content with loader:', LoaderClass.name);
        aggregatedContent = await loaderInstance.aggregateContent(pages);
        log('Content aggregated successfully, length:', aggregatedContent.length);
      } catch (aggError) {
        const error = aggError as Error;
        console.error(
          `Error aggregating content for ${filePath} using ${LoaderClass.name}: ${error.message}`,
        );
        aggregationError = `Content aggregation failed: ${error.message}`;
        // Keep the pages loaded, but content might be empty/incomplete
      }

      // 3. Attach document-specific metadata if loader supports it
      if (typeof loaderInstance.attachDocumentMetadata === 'function') {
        log('Loader supports attachDocumentMetadata. Attaching...');
        try {
          loaderSpecificMetadata = await loaderInstance.attachDocumentMetadata(filePath);
          log('Document-specific metadata attached:', loaderSpecificMetadata);
        } catch (metaErr) {
          const error = metaErr as Error;
          console.error(
            `Error attaching metadata for ${filePath} using ${LoaderClass.name}: ${error.message}`,
          );
          metadataError = `Metadata attachment failed: ${error.message}`;
        }
      } else {
        log('Loader does not support attachDocumentMetadata.');
      }
    } catch (loadErr) {
      const error = loadErr as Error;
      console.error(
        `Error loading pages for ${filePath} using ${LoaderClass.name}: ${error.message}`,
      );
      loaderError = `Loader execution failed: ${error.message}`;
      // Provide a minimal error page if loader failed critically
      pages = [
        {
          charCount: 0,
          lineCount: 0,
          metadata: { error: loaderError },
          pageContent: '',
        },
      ];
      // Aggregated content remains empty
    }
  } else {
    log('File stats access failed (fsError: %s). Creating minimal error page.', fsError);
    // If stats failed, create a minimal error page
    pages = [
      {
        charCount: 0,
        lineCount: 0,
        metadata: { error: fsError },
        pageContent: '',
      },
    ];
    // Aggregated content remains empty
  }

  // Calculate totals from the loaded pages
  let totalCharCount = 0;
  let totalLineCount = 0;
  log('Calculating total char and line counts from pages.');
  for (const page of pages) {
    totalCharCount += page.charCount;
    totalLineCount += page.lineCount;
  }
  log('Totals calculated:', { totalCharCount, totalLineCount });

  // Combine all potential errors
  const combinedError =
    [fsError, loaderError, aggregationError, metadataError].filter(Boolean).join('; ') || undefined;
  if (combinedError) log('Combined errors:', combinedError);

  // Construct the final FileDocument
  log('Constructing final FileDocument.');
  const fileDocument: FileDocument = {
    content: aggregatedContent, // Use content from aggregateContent
    createdTime,
    fileType,
    filename,
    metadata: {
      // Include combined errors
      error: combinedError,
      // Add loader specific metadata under a namespace
      loaderSpecific: loaderSpecificMetadata ?? undefined,
      // Add other file-level metadata
      ...fileMetadata,
    },
    modifiedTime,
    pages, // Use pages from loadPages
    source,
    totalCharCount,
    totalLineCount,
  };

  // Clean up undefined error field if no error occurred
  if (!fileDocument.metadata.error) {
    delete fileDocument.metadata.error;
  }

  log('File loading process completed for:', filePath, 'Returning document:', {
    fileType: fileDocument.fileType,
    filename: fileDocument.filename,
    pages: fileDocument.pages?.length,
  });
  return fileDocument;
};
