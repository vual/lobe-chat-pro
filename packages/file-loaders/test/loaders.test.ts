// @vitest-environment node
import fs from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';

import { loadFile } from '../src';

const getFixturePath = (filename: string) => path.join(__dirname, 'fixtures', filename);

const TEXT_FILES = ['test.txt', 'test.csv', 'test.md'];

describe('loadFile Integration Tests', () => {
  describe('Text Handling (.txt, .csv, .md, etc.)', () => {
    const testPureTextFile = (fileName: string) => {
      it(`should load content from a ${fileName} file using filePath`, async () => {
        const filePath = getFixturePath(fileName);
        const expectedContent = fs.readFileSync(filePath, 'utf-8');

        // Pass filePath directly to loadFile
        const docs = await loadFile(filePath);

        expect(docs.content).toEqual(expectedContent);
        expect(docs.source).toEqual(filePath);

        // @ts-expect-error
        delete docs.source;
        // @ts-expect-error
        delete docs.createdTime;
        // @ts-expect-error
        delete docs.modifiedTime;
        expect(docs).toMatchSnapshot();
      });
    };

    TEXT_FILES.forEach((file) => {
      testPureTextFile(file);
    });
  });

  describe('PDF Handling', () => {
    it(`should load content from a pdf file using filePath`, async () => {
      const filePath = getFixturePath('test.pdf');

      // Pass filePath directly to loadFile
      const docs = await loadFile(filePath);

      expect(docs.content).toContain('123');
      expect(docs.source).toEqual(filePath);

      // @ts-expect-error
      delete docs.source;
      // @ts-expect-error
      delete docs.createdTime;
      // @ts-expect-error
      delete docs.modifiedTime;
      expect(docs).toMatchSnapshot();
    });
  });
});
