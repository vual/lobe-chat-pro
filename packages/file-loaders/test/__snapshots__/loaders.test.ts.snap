// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadFile Integration Tests > PDF Handling > should load content from a pdf file using filePath 1`] = `
{
  "content": "<page pageNumber="1">
123
</page>",
  "fileType": "pdf",
  "filename": "test.pdf",
  "metadata": {
    "loaderSpecific": {
      "pdfInfo": {
        "CreationDate": "D:20250419143655Z00'00'",
        "Creator": "Pages文稿",
        "EncryptFilterName": null,
        "IsAcroFormPresent": false,
        "IsCollectionPresent": false,
        "IsLinearized": false,
        "IsSignaturesPresent": false,
        "IsXFAPresent": false,
        "Language": null,
        "ModDate": "D:20250419143655Z00'00'",
        "PDFFormatVersion": "1.3",
        "Producer": "macOS 版本15.3.2（版号24D81） Quartz PDFContext",
        "Title": "test",
      },
      "pdfMetadata": null,
      "pdfVersion": "4.10.38",
    },
  },
  "pages": [
    {
      "charCount": 3,
      "lineCount": 1,
      "metadata": {
        "pageNumber": 1,
      },
      "pageContent": "123",
    },
  ],
  "totalCharCount": 3,
  "totalLineCount": 1,
}
`;

exports[`loadFile Integration Tests > Text Handling (.txt, .csv, .md, etc.) > should load content from a test.csv file using filePath 1`] = `
{
  "content": "ID,Name,Value
1,Alpha,100
2,Beta,200
3,Gamma,300
",
  "fileType": "csv",
  "filename": "test.csv",
  "metadata": {
    "loaderSpecific": undefined,
  },
  "pages": [
    {
      "charCount": 49,
      "lineCount": 5,
      "metadata": {
        "lineNumberEnd": 5,
        "lineNumberStart": 1,
      },
      "pageContent": "ID,Name,Value
1,Alpha,100
2,Beta,200
3,Gamma,300
",
    },
  ],
  "totalCharCount": 49,
  "totalLineCount": 5,
}
`;

exports[`loadFile Integration Tests > Text Handling (.txt, .csv, .md, etc.) > should load content from a test.md file using filePath 1`] = `
{
  "content": "# Markdown Test

This is a test.
",
  "fileType": "md",
  "filename": "test.md",
  "metadata": {
    "loaderSpecific": undefined,
  },
  "pages": [
    {
      "charCount": 33,
      "lineCount": 4,
      "metadata": {
        "lineNumberEnd": 4,
        "lineNumberStart": 1,
      },
      "pageContent": "# Markdown Test

This is a test.
",
    },
  ],
  "totalCharCount": 33,
  "totalLineCount": 4,
}
`;

exports[`loadFile Integration Tests > Text Handling (.txt, .csv, .md, etc.) > should load content from a test.txt file using filePath 1`] = `
{
  "content": "This is line 1.
This is line 2.
End of text file.
",
  "fileType": "txt",
  "filename": "test.txt",
  "metadata": {
    "loaderSpecific": undefined,
  },
  "pages": [
    {
      "charCount": 50,
      "lineCount": 4,
      "metadata": {
        "lineNumberEnd": 4,
        "lineNumberStart": 1,
      },
      "pageContent": "This is line 1.
This is line 2.
End of text file.
",
    },
  ],
  "totalCharCount": 50,
  "totalLineCount": 4,
}
`;
